packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -157,6 +170,56 @@ class NoDeltaStream
     public close(): void { this.dispose(); }
 }
 
+/**
+ * This class exists to help diagnose issues described in https://github.com/microsoft/FluidFramework/issues/7312
+ *
+ * This adapter class tracks resume & pause calls to provide ability to debug cases where
+ * someone is holding queue in paused state too long.
+ * The only thing that it does on top of DeltaQueue is tracking all users who called pause()
+ * and making available callstacks for those calls.
+ *
+ * Long term, we better off switching to tracking callers by
+ *  - pause() returning token that is required to be passed into resume()
+ *  - tracking tokens and stacks with tokens
+ * This would be improvement over current system, as today we can't pair pause & resume calls.
+ */
+class DeltaQueueProxy<T> extends EventForwarder<IDeltaQueueEvents<T>> implements IDeltaQueue<T>, IDeltaQueueWriter<T> {
+    // IDeltaQueueWriter methods
+    public push(task: T) { return this.queue.push(task); }
+    public clear()  { return this.queue.clear(); }
+
+    // IDeltaQueue methods
+    public get paused() { return this.queue.paused; }
+    public get length() { return this.queue.length; }
+    public get idle() { return this.queue.idle; }
+    public peek() { return this.queue.peek(); }
+    public toArray() { return this.queue.toArray(); }
+    public async waitTillProcessingDone() { return this.queue.waitTillProcessingDone(); }
+
+    public async pause() {
+        // Note: this is not cheap. On my rather fast machine it takes ~0.08ms per call
+        // If this becomes perf problem, we need to redesign how often we pause op processing -
+        // it just not right to call it for every incoming op!
+        this.stacks.push(generateStack() ?? "");
```

Comment:
We should be able to optimize `generateStack`.  
I believe you don't need to call `.stack` right away and just save the Error object.  Measurement on node is about 6x slower with the `.stack` call.

There might be browser that behave differently, but we can do a one time probe to check the browser behavior and behave different on that (instead of checking every time.

Reply:
Good point!

Reply:
We should do this for #7394